crates/paimon/src/file_index/file_index_format.rs

```diff
@@ -0,0 +1,514 @@
+                current_offset += index_name_len as u64;
+
+                // Start Pos (4 bytes)
+                buffer = self.reader.read(current_offset..current_offset + 4).await?;
+                let start_pos = buffer.get_i32_le();
+                current_offset += 4;
+
+                // Length (4 bytes)
+                buffer = self.reader.read(current_offset..current_offset + 4).await?;
+                let length = buffer.get_i32_le();
+                current_offset += 4;
+
+                index_info_map.insert(index_name, IndexInfo { start_pos, length });
+            }
+
+            self.header.insert(column_name, index_info_map);
+        }
+
+        buffer = self.reader.read(current_offset..current_offset + 4).await?;
+        let redundant_length = buffer.get_i32_le() as u64;
+        current_offset += 4;
+
+        if redundant_length > 0 {
+            let redundant_bytes = self
+                .reader
+                .read(current_offset..current_offset + redundant_length)
+                .await?;
+
+            if redundant_bytes.len() as u64 != redundant_length {
+                return Err(Error::FileIndexFormatInvalid {
+                    message: format!(
+                        "Expected to read {} redundant bytes, but found only {}",
+                        redundant_length,
+                        redundant_bytes.len()
+                    ),
+                });
+            }
+        }
+
+        Ok(())
+    }
+
+    pub async fn read_column_index(
+        &self,
+        column_name: &str,
+    ) -> crate::Result<HashMap<String, Bytes>> {
+        if let Some(index_info) = self.header.get(column_name) {
+            let mut result = HashMap::new();
+            for (index_name, info) in index_info {
+                let bytes = self.get_bytes_with_start_and_length(info).await?;
+                result.insert(index_name.clone(), bytes);
+            }
+            Ok(result)
+        } else {
+            Err(Error::FileIndexFormatInvalid {
+                message: format!("Column '{}' not found in header", column_name),
+            })
+        }
+    }
+
+    async fn get_bytes_with_start_and_length(
+        &self,
+        index_info: &IndexInfo,
+    ) -> crate::Result<Bytes> {
+        let data_bytes = self
+            .reader
+            .read(index_info.start_pos as u64..(index_info.start_pos + index_info.length) as u64)
+            .await?;
+
+        Ok(data_bytes)
+    }
+
+    pub async fn from_file_io(file_io: FileIO, path: &str) -> crate::Result<Self> {
+        let input_file = file_io.new_input(path)?;
+        Self::new(input_file).await
+    }
+}
+
+#[cfg(test)]
+mod file_index_format_tests {
+    use super::*;
+    use bytes::Bytes;
+    use std::collections::HashMap;
+
+    #[tokio::test]
+    async fn test_single_column_single_index() -> crate::Result<()> {
+        let path = "file:/tmp/test_single_column_single_index";
```

Comment:
We can use in-memory file I/O for testing.

Reply:
No problem.